### PR TITLE
Switched to the new Swift native JSON type for --dump-package

### PR DESCRIPTION
**Pull request by czechboy0**

### CODE REVIEW

**Comment by ddunbar** [comment]
Do you care about losing the print printing, or want to merge and get that later?

**Comment by ddunbar** [comment]
Linux fails:
```
/home/ddunbar/public/swift-project/swiftpm/Sources/PackageLoading/JSONSerialization.swift:22:42: error: type 'NSJSONWritingOptions' has no member 'prettyPrinted'                                                                   
    let options: NSJSONWritingOptions = .prettyPrinted                                                                                                                                                                              
                                         ^~~~~~~~~~~~~                                                                                                                                                                              
```

File: Sources/PackageLoading/JSONSerialization.swift [hunk 1]
```diff
@@ -8,115 +8,78 @@
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
  */
 
-import Foundation
 import PackageDescription
+import Basic
```
Comment (ddunbar) [inline Sources/PackageLoading/JSONSerialization.swift]: Nit: Can we keep imports lexicographically ordered?

File: Sources/PackageLoading/JSONSerialization.swift [hunk 2]
```diff
@@ -8,115 +8,78 @@
...
 
 /// A JSON representation of an element.
 protocol JSONSerializable {
     
     /// Return a JSON representation.
-    func toJSON() -> AnyObject
+    func toJSON() -> JSON
 }
```
Comment (ddunbar) [inline Sources/PackageLoading/JSONSerialization.swift]: Let's just move this protocol to Basic.JSON?

File: Sources/PackageLoading/JSONSerialization.swift [hunk 3]
```diff
@@ -8,115 +8,78 @@
...
-extension Array {
-    func asNS() -> NSArray {
-        #if os(Linux)
-            return self.bridge()
-        #else
-            return self.map { $0 as! AnyObject } as NSArray
-        #endif
+    let json = package.toJSON()
+    guard let string = json.toBytes().asString else {
+        fatalError("Failed to serialize JSON \(json)")
```
Comment (ddunbar) [inline Sources/PackageLoading/JSONSerialization.swift]: This shouldn't actually ever be possible (unless there was a bug in the JSON formatter in encoding bytes), so what I think we should get a `JSON.toString()` method that does it and doesn't fail.

**Comment by ddunbar** [comment]
Patch LGTM, I added a few comments that you can either address along with this or in a separate PR if you prefer.

File: Sources/PackageLoading/JSONSerialization.swift (hunk 1 above)
Comment (czechboy0) [inline Sources/PackageLoading/JSONSerialization.swift]: Sure thing

File: Sources/PackageLoading/JSONSerialization.swift (hunk 2 above)
Comment (czechboy0) [inline Sources/PackageLoading/JSONSerialization.swift]: Okay

**Comment by czechboy0** [comment]
@ddunbar Updated, I also created a `JSONConvertible` protocol, which allows non-collection types to be wrapped implicitly (so instead of `.string("foo")` you just need `"foo"`). Can't do this for collections until it's possible to add conformance of a constrained array/dictionary.

File: Sources/Basic/JSON.swift [hunk 4]
```diff
@@ -39,10 +39,17 @@ public enum JSON {
     case string(String)
 
     /// An array.
-    case array([JSON])
+    case array([JSONConvertible])
 
     /// A dictionary.
-    case dictionary([String: JSON])
+    case dictionary([String: JSONConvertible])
```
Comment (ddunbar) [inline Sources/Basic/JSON.swift]: I would rather not change this... arrays of protocol type are going to be more expensive than arrays of a known fixed type.

File: Sources/Basic/JSON.swift [hunk 5]
```diff
@@ -39,10 +39,17 @@ public enum JSON {
...
+}
+
+/// A JSON representation of an element.
+public protocol JSONConvertible {
+    
+    /// Return a JSON representation.
+    var json: JSON { get }
```
Comment (ddunbar) [inline Sources/Basic/JSON.swift]: I actually preferred this as a function -- I don't think properties should be used when they aren't close to O(1), and I would expect most implementations to just perform the transformation.

File: Sources/PackageLoading/JSONSerialization.swift (hunk 3 above)
Comment (czechboy0) [inline Sources/PackageLoading/JSONSerialization.swift]: Fixed

**Comment by czechboy0** [comment]
@ddunbar 🔔

**Comment by ddunbar** [comment]
LGTM

**Pull request merged**